src/adc_readout/AdcTimeStamp.cpp

```diff
@@ -11,50 +11,79 @@
 
 const static std::uint64_t NSecMultiplier = 1000000000;
 
-RawTimeStamp::RawTimeStamp(std::uint64_t NSec)
-    : Seconds(NSec / NSecMultiplier) {
+struct Timing {
+  constexpr Timing(std::uint64_t const ClockFrequency)
+      : TimeClockFrequency(ClockFrequency), SamplingRate(ClockFrequency / 2),
+        AdcTimerCounterMax(SamplingRate), SampleLengthNS(2e9 / ClockFrequency) {
+  }
+  std::uint32_t const TimeClockFrequency;
+  std::uint32_t const SamplingRate;
+  std::uint32_t const AdcTimerCounterMax;
+  double const SampleLengthNS;
+};
+
+static constexpr Timing const TimeConst[2] = {TimerClockFrequencyInternal,
+                                              TimerClockFrequencyExternal};
+
+TimeStamp::TimeStamp(std::uint64_t NSec, ClockMode Mode) : CMode(Mode) {
+  CTime.Seconds = NSec / NSecMultiplier;
   auto NanoSecPart = NSec % NSecMultiplier;
-  SecondsFrac = std::lround(NanoSecPart / SampleLengthNS);
-  if (SecondsFrac == AdcTimerCounterMax) {
-    ++Seconds;
-    SecondsFrac = 0;
+  auto &CConst = TimeConst[int(CMode)];
+  CTime.SecondsFrac =
+      std::lround(NanoSecPart / TimeConst[int(CMode)].SampleLengthNS);
+  if (CTime.SecondsFrac == CConst.AdcTimerCounterMax) {
+    ++CTime.Seconds;
+    CTime.SecondsFrac = 0;
   }
 }
 
-std::uint64_t RawTimeStamp::getTimeStampNS() const {
-  auto NanoSec = static_cast<std::uint64_t>(
-      std::llround(static_cast<double>(SecondsFrac) /
-                   static_cast<double>(AdcTimerCounterMax) * 1e9));
+std::uint64_t TimeStamp::getTimeStampNS() const {
+  auto NanoSec = static_cast<std::uint64_t>(std::llround(
+      static_cast<double>(CTime.SecondsFrac) /
+      static_cast<double>(TimeConst[int(CMode)].AdcTimerCounterMax) * 1e9));
   return static_cast<std::uint64_t>(
-      static_cast<std::uint64_t>(Seconds) * NSecMultiplier + NanoSec);
+      static_cast<std::uint64_t>(CTime.Seconds) * NSecMultiplier + NanoSec);
 }
 
 // Note: This function might be significantly slower than CalcTimeStamp() for
 // some cases.
-std::uint64_t RawTimeStamp::getTimeStampNSFast() const {
+std::uint64_t TimeStamp::getTimeStampNSFast() const {
 
-  std::uint64_t NanoSec =
-      (((SecondsFrac * 100000000000) / (AdcTimerCounterMax)) + 50) / 100;
+  std::uint64_t NanoSec = (((CTime.SecondsFrac * 100000000000) /
```

Comment:
A formula like this should maybe  have some explanatory text. 
Also what does the hardcoded 50 and 100 signify?

Reply:
This is a fixed point calculation of a decimal value, hence 50 corresponds to 0.5 and 100 makes it such that this is the case. Doing fixed point math can be significantly faster than floating point math but based on my benchmarks this is not necessarily the case on modern Intel CPUs that have powerful floating point units. I will hence remove the code.